File: examples/dsl-discount-spec/app/discount.js
// discount.js — the business rule, on its own so both the checker and the app share it.
//
// Part 1 of the workshop builds this file plus the runner that checks it.
// Part 2's browser app (index.html + app.js) imports this same function for its UI, so the
// thing the spec checks and the thing the user clicks are the same code — no second copy to drift.
//
// The exact rules are defined in ../specs/discount.feature. This file must satisfy them;
// don't edit the spec to match the code.

function apply_discount(cartTotal, code) {
  const RULES = {
    SAVE10: { rate: 0.1, min: 50 }, // 10% off, only if cart_total >= 50
    SAVE20: { rate: 0.2, min: 100 }, // 20% off, only if cart_total >= 100
  };

  const rule = RULES[code];
  if (!rule) {
    return { applied: false, final_total: cartTotal, reason: "invalid code" };
  }
  if (cartTotal < rule.min) {
    return { applied: false, final_total: cartTotal, reason: "minimum not met" };
  }

  // Round to cents so 120.00 * 0.8 lands exactly on 96.00.
  const final_total = Math.round(cartTotal * (1 - rule.rate) * 100) / 100;
  return { applied: true, final_total, reason: `${code} applied` };
}

// Node (the runner) imports it; the browser reads it as a global from <script src="discount.js">.
if (typeof module !== "undefined" && module.exports) {
  module.exports = { apply_discount };
}

File: examples/dsl-discount-spec/discount.py
"""discount.py — the business rule, imported by dsl_runner.py.

Faithful Python twin of app/discount.js's apply_discount (same rules, same field names), so the
Python runner (python dsl_runner.py specs/discount.feature) checks the exact same behavior the JS
runner does. The exact rules are defined in specs/discount.feature; this file must satisfy them —
don't edit the spec to match the code.
"""

# 10% off only if cart_total >= 50; 20% off only if cart_total >= 100. Pinned by the boundary
# scenarios in specs/discount.feature (t-ef19) — not left to inference.
RULES = {
    "SAVE10": {"rate": 0.1, "min": 50},
    "SAVE20": {"rate": 0.2, "min": 100},
}


def apply_discount(cart_total: float, code: str) -> dict:
    rule = RULES.get(code)
    if rule is None:
        return {"applied": False, "final_total": cart_total, "reason": "invalid code"}
    if cart_total < rule["min"]:
        return {"applied": False, "final_total": cart_total, "reason": "minimum not met"}
    # Round to cents so 120.00 * 0.8 lands exactly on 96.00.
    final_total = round(cart_total * (1 - rule["rate"]) + 1e-9, 2)
    return {"applied": True, "final_total": final_total, "reason": f"{code} applied"}
